harness/tests/experiment/fixtures/pytorch_onevar_model.py

```diff
@@ -221,12 +227,152 @@ def build_validation_data_loader(self) -> torch.utils.data.DataLoader:
+
     def __init__(self, context: pytorch.PyTorchTrialContext) -> None:
         super().__init__(context)
         self.model, self.optimizer = self.context.configure_apex_amp(
-            models=self.model, optimizers=self.opt
+            models=self.model,
+            optimizers=self.opt,
+            opt_level="O1",
+        )
+
+    def train_batch(
+        self, batch: pytorch.TorchData, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+        scale_before = apex.amp.state_dict()["loss_scaler0"]["loss_scale"]
+        metrics = super().train_batch(batch, epoch_idx, batch_idx)
+        metrics["scale_before"] = scale_before
+        metrics["scale"] = apex.amp.state_dict()["loss_scaler0"]["loss_scale"]
+        metrics["stage"] = self._dataset_args[0][batch_idx]
+        return metrics
+
+
+class OneVarAutoAMPTrial(OneVarAMPBaseTrial):
+    _init_scale = 65536
+    _growth_interval = 4
+
+    def __init__(self, context: pytorch.PyTorchTrialContext) -> None:
+        context.experimental.use_amp(
+            gradscaler_kwargs={
+                "init_scale": self._init_scale,
+                "growth_interval": self._growth_interval,
+            }
         )
+        super().__init__(context)
+        self.scaler = self.context._scaler
+
+    def train_batch(
+        self, batch: pytorch.TorchData, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+        scale_before = self.scaler.get_scale()
+        metrics = super().train_batch(batch, epoch_idx, batch_idx)
+        metrics["scale_before"] = scale_before
+        # self.scaler.update() gets called after this method returns
+        metrics["stage"] = self._dataset_args[0][batch_idx]
+        return metrics
+
+
+class OneVarManualAMPTrial(OneVarAMPBaseTrial):
+    _init_scale = 65536
+    _growth_interval = 4
+
+    def __init__(self, context: pytorch.PyTorchTrialContext) -> None:
+        self.scaler = context.wrap_scaler(
+            torch.cuda.amp.GradScaler(
+                init_scale=self._init_scale, growth_interval=self._growth_interval
+            )
+        )
+        super().__init__(context)
+
+    def train_batch(
+        self, batch: pytorch.TorchData, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+        data, label = batch
+
+        scale_before = self.scaler.get_scale()
+
+        # Measure the weight right now.
+        w_before = self.model.weight.data.item()
+
+        # Calculate expected values for loss (eq 1) and weight (eq 4).
+        loss_exp = (label[0] - data[0] * w_before) ** 2
+        w_exp = w_before + 2 * self.lr * data[0] * (label[0] - (data[0] * w_before))
+
+        with torch.cuda.amp.autocast():
+            output = self.model(data)
+            loss = self.loss_fn(output, label)
+
+        scaled_loss = self.scaler.scale(loss)
+        self.context.backward(scaled_loss)
+        self.context.step_optimizer(self.opt, scaler=self.scaler)
+        self.scaler.update()
```

Comment:
can you write the test without this?  It's technically not allowed in the PyTorchTrial API, and so the times we call scaler.update() are going to cause the control loop to run at the wrong rate.

Reply:
i think this is the manual case, in which the auto_amp scaler.update() doesn't actually get called, so we do need this.

Reply:
oh... duh.  I missed the `or not self.context.experimental._auto_amp` in `_should_update_scaler`.  Thank you.